### ISSUE TITLE
*: make migration state testing trivial

### ISSUE THREAD

**Issue by ajwerner** [cockroachdb/cockroach#72983]
**Is your feature request related to a problem? Please describe.**
We do an incredibly bad job testing clusters which are using the new binary but have state that was bootstrapped in an older version and has either not been migrated or has been partially migrated. Currently there is no way to do this from a unit test.

**Describe the solution you'd like**

* Provide a mechanism to bootstrap test clusters from store dir state that is checked in.
    * This should be easy, just provide a mechanism to initialize the in-memory filesystem from a real directory
* Provide tooling to generate store dirs by downloading old binaries and running upgrade.
* Extend the above tool to pause in the middle of various migrations and snapshot those directories too.
* Hook into `testcluster.StartTestCluster` and `testserver.StartTestServer` to use these store dirs
    * Lots of questions here regarding the exact semantics.
    * Something cool would be to randomize the initial cluster state in most cases.
    * Additionally we'll want to add explicit selection of the state
    * Logictest configs seem obvious.
    
Some care is going to need to be taken to make sure we properly maintain and augment these store dirs when we cut releases.

**Describe alternatives you've considered**
This won't help us detect any binary incompatibilities in the mixed version state between nodes of the old version and nodes of the new version. To deal with that, we could imagine actually downloading and running the old version binaries and making them part of the cluster. It might be nice to leverage [testserver](https://pkg.go.dev/github.com/cockroachdb/cockroach-go/v2/testserver) for that. This would probably be way slower and have many more concerns from a developer experience perspective. Nevertheless, it'll probably be useful and we should be dogfooding that library more.

**Additional context**

For the automation creating the store dirs, it might be nice to leverage [testserver](https://pkg.go.dev/github.com/cockroachdb/cockroach-go/v2/testserver).

**Comment by miretskiy** [comment]
I think these are good ideas; perhaps more applicable to roachtests than unit tests.
Though, say a metamorphic constant that chooses to bootstrap new cluster with *small* pre-built state, might be pretty nice, as long as these migrations are observable and debuggable if some unrelated test fails (unrelated in a sense that it's not a migration test, but some unit tests that simply tried to start the server).

**Comment by ajwerner** [comment]
>I think these are good ideas; perhaps more applicable to roachtests than unit tests.

I strongly disagree. If we're going roachtests, then we can spin up real clusters. 

----

I think you're maybe missing the proposal. I'm saying every single unit test which uses a test cluster or test server, by default, should not get to assume that it was bootstrapped at the current version.
